third/nmh/man/rcvdist.man
.\"
.\" %nmhwarning%
.\" $Id: rcvdist.man,v 1.1.1.1 1999-02-07 18:14:22 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH RCVDIST %manext1% MH.6.8 [%nmhversion%]
.SH NAME
rcvdist \- asynchronously redistribute new mail
.SH SYNOPSIS
.in +.5i
.ti -.5i
%libdir%/rcvdist
\%[\-form\ formfile]
.br
\%[switches\ for\ \fIpostproc\fR]
address1\ ...
.br
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
The \fIrcvdist\fR program will accept a message on its standard input
and resend a copy of this message to all of the addresses listed on its
command line.

When a message is redistributed with the \fIrcvdist\fR command, the
format of the Resent-xxx header fields is controlled by the forms files
"rcvdistcomps".  If a file named "rcvdistcomps" exists in the user's nmh
directory, it will be used instead of the default one.  You may specify
an alternate forms file with the switch `\-form\ formfile'.

The "rcvdistcomps" file uses the format string facility described in
\fImh\-format\fR(5).  In addition to the standard format escapes,
\fIrcvdist\fP also recognizes the following additional \fIcomponent\fR
escape:
.sp 1
.ne 5
.nf
.ta \w'Dtimenow  'u +\w'Returns  'u
\fIEscape\fR     \fIReturns\fR   \fIDescription\fR
addresses  string    the addresses to distribute to
.re
.fi

By default, \fIrcvdist\fR uses the program \fIpost\fR(8) to do the actual
delivery of the message, although this can be changed by defining the
\fIpostproc\fR profile component.
.Fi
^%etcdir%/rcvdistcomps~^Default message skeleton
^or <mh\-dir>/rcvdistcomps~^Rather than standard message skeleton
^%etcdir%/mts.conf~^nmh mts configuration file
^$HOME/\&.maildelivery~^The file controlling local delivery
^%etcdir%/maildelivery~^Rather than the standard file
.Sa
rcvpack(1), rcvstore(1), rcvtty(1), mh\-format(5), slocal(1)
.Bu
Only two return codes are meaningful, others should be.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/rcvdist.c

/*
 * rcvdist.c -- asynchronously redistribute messages
 *
 * $Id: rcvdist.c,v 1.1.1.1 1999-02-07 18:14:16 danw Exp $
 */

#include <h/mh.h>
#include <h/fmt_scan.h>
#include <h/rcvmail.h>
#include <zotnet/tws/tws.h>

static struct swit switches[] = {
#define	FORMSW       0
    { "form formfile",  4 },
#define VERSIONSW    1
    { "version", 0 },
#define	HELPSW       2
    { "help", 4 },
    { NULL, 0 }
};

static char backup[BUFSIZ] = "";
static char drft[BUFSIZ] = "";
static char tmpfil[BUFSIZ] = "";

/*
 * prototypes
 */
static void rcvdistout (FILE *, char *, char *);
void done (int);


int
main (int argc, char **argv)
{
    pid_t child_id;
    int i, vecp = 1;
    char *addrs = NULL, *cp, *form = NULL, buf[BUFSIZ];
    char **argp, **arguments, *vec[MAXARGS];
    register FILE *fp;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    mts_init (invo_name);
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    vec[vecp++] = --cp;
		    continue;

		case HELPSW: 
		    snprintf (buf, sizeof(buf),
			"%s [switches] [switches for postproc] address ...",
			invo_name);
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;
	    }
	}
	addrs = addrs ? add (cp, add (", ", addrs)) : getcpy (cp);
    }

    if (addrs == NULL)
	adios (NULL, "usage: %s [switches] [switches for postproc] address ...",
	    invo_name);

    umask (~m_gmprot ());
    strncpy (tmpfil, m_tmpfil (invo_name), sizeof(tmpfil));
    if ((fp = fopen (tmpfil, "w+")) == NULL)
	adios (tmpfil, "unable to create");
    cpydata (fileno (stdin), fileno (fp), "message", tmpfil);
    fseek (fp, 0L, SEEK_SET);
    strncpy (drft, m_tmpfil (invo_name), sizeof(drft));
    rcvdistout (fp, form, addrs);
    fclose (fp);

    if (distout (drft, tmpfil, backup) == NOTOK)
	done (1);

    vec[0] = r1bindex (postproc, '/');
    vec[vecp++] = "-dist";
    vec[vecp++] = drft;
    vec[vecp] = NULL;

    for (i = 0; (child_id = fork()) == NOTOK && i < 5; i++)
	sleep (5);
    switch (child_id) {
	case NOTOK: 
	    admonish (NULL, "unable to fork");/* fall */
	case OK: 
	    execvp (postproc, vec);
	    fprintf (stderr, "unable to exec ");
	    perror (postproc);
	    _exit (1);

	default: 
	    done (pidXwait(child_id, postproc));
    }
/* NOTREACHED */
}

/* very similar to routine in replsbr.c */

#define	SBUFSIZ	256

static int outputlinelen = OUTPUTLINELEN;

static struct format *fmt;

static int ncomps = 0;
static char **compbuffers = 0;
static struct comp **used_buf = 0;

static int dat[5];

static char *addrcomps[] = {
    "from",
    "sender",
    "reply-to",
    "to",
    "cc",
    "bcc",
    "resent-from",
    "resent-sender",
    "resent-reply-to",
    "resent-to",
    "resent-cc",
    "resent-bcc",
    NULL
};


static void
rcvdistout (FILE *inb, char *form, char *addrs)
{
    register int char_read = 0, format_len, i, state;
    register char *tmpbuf, **nxtbuf, **ap;
    char *cp, *scanl, name[NAMESZ];
    register struct comp *cptr, **savecomp;
    FILE *out;

    if (!(out = fopen (drft, "w")))
	adios (drft, "unable to create");

    /* get new format string */
    cp = new_fs (form ? form : rcvdistcomps, NULL, NULL);
    format_len = strlen (cp);
    ncomps = fmt_compile (cp, &fmt) + 1;
    if (!(nxtbuf = compbuffers = (char **) calloc ((size_t) ncomps, sizeof(char *))))
	adios (NULL, "unable to allocate component buffers");
    if (!(savecomp = used_buf = (struct comp **) calloc ((size_t) (ncomps + 1), sizeof(struct comp *))))
	adios (NULL, "unable to allocate component buffer stack");
    savecomp += ncomps + 1;
    *--savecomp = 0;

    for (i = ncomps; i--;)
	if (!(*nxtbuf++ = malloc (SBUFSIZ)))
	    adios (NULL, "unable to allocate component buffer");
    nxtbuf = compbuffers;
    tmpbuf = *nxtbuf++;

    for (ap = addrcomps; *ap; ap++) {
	FINDCOMP (cptr, *ap);
	if (cptr)
	    cptr->c_type |= CT_ADDR;
    }

    FINDCOMP (cptr, "addresses");
    if (cptr)
	cptr->c_text = addrs;

    for (state = FLD;;) {
	switch (state = m_getfld (state, name, tmpbuf, SBUFSIZ, inb)) {
	    case FLD: 
	    case FLDPLUS: 
		if ((cptr = wantcomp[CHASH (name)]))
		    do {
			if (!strcasecmp (name, cptr->c_name)) {
			    char_read += msg_count;
			    if (!cptr->c_text) {
				cptr->c_text = tmpbuf;
				*--savecomp = cptr;
				tmpbuf = *nxtbuf++;
			    }
			    else {
				i = strlen (cp = cptr->c_text) - 1;
				if (cp[i] == '\n')
				    if (cptr->c_type & CT_ADDR) {
					cp[i] = 0;
					cp = add (",\n\t", cp);
				    }
				    else
					cp = add ("\t", cp);
				cptr->c_text = add (tmpbuf, cp);
			    }
			    while (state == FLDPLUS) {
				state = m_getfld (state, name, tmpbuf,
						  SBUFSIZ, inb);
				cptr->c_text = add (tmpbuf, cptr->c_text);
				char_read += msg_count;
			    }
			    break;
			}
		    } while ((cptr = cptr->c_next));

		while (state == FLDPLUS)
		    state = m_getfld (state, name, tmpbuf, SBUFSIZ, inb);
		break;

	    case LENERR: 
	    case FMTERR: 
	    case BODY: 
	    case FILEEOF: 
		goto finished;

	    default: 
		adios (NULL, "m_getfld() returned %d", state);
	}
    }
finished: ;

    i = format_len + char_read + 256;
    scanl = malloc ((size_t) i + 2);
    dat[0] = dat[1] = dat[2] = dat[4] = 0;
    dat[3] = outputlinelen;
    fmt_scan (fmt, scanl, i, dat);
    fputs (scanl, out);

    if (ferror (out))
	adios (drft, "error writing");
    fclose (out);

    free (scanl);
    for (nxtbuf = compbuffers, i = ncomps; cptr = *savecomp++; nxtbuf++, i--)
	free (cptr->c_text);
    while (i-- > 0)
        free (*nxtbuf++);
    free ((char *) compbuffers);
    free ((char *) used_buf);
}


void
done (int status)
{
    if (backup[0])
	unlink (backup);
    if (drft[0])
	unlink (drft);
    if (tmpfil[0])
	unlink (tmpfil);

    exit (status ? RCV_MBX : RCV_MOK);
}
